### PR TITLE
Fake PR to give a place for comments on entire repo

### PR DESCRIPTION
**Pull request by jutley**

### CODE REVIEW

File: kafka-broker-cfn.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
+AWSTemplateFormatVersion: '2010-09-09'
+Description: per-az resources for kafka
+
+Parameters:
+  BaseStack:
+    Type: String
+    Default: hack
```
Comment (jbarrioshiya) [inline kafka-broker-cfn.yaml]: Fake comment on this fake PR

File: bin/bind [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (kiyose) [inline bin/bind]: I'm not sure what you mean when you call this file "bind". To me bind is a DNS server.  The file needs to have a clear name and should really have a help that describes what it does.

File: bin/bind [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+##   -in-file string    Path to the container linux config. Standard input unless specified otherwise.
+##   -out-file string   Path to the resulting Ignition config. Standard output unless specified otherwise.
+##   -platform string   Platform to target. Accepted values: [azure digitalocean ec2 gce packet openstack-metadata vagrant-virtualbox cloudstack-configdrive custom].
+##   -pretty            Indent the resulting Ignition config.
+##   -strict            Fail if any warnings are encountered.
+##   -version           Print the version and exit.
+
+
+export REGION=us-west-2
+pushd $ROOT
```
Comment (akunszt) [inline bin/bind]: Please quote your variables: `pushd "$ROOT"`

File: bin/bind [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+mkdir -p target
+for i in cl-config/*.yaml; do
+  echo "Transpiling $i ======================"
+  export CLASS=$(basename ${i%.yaml})
```
Comment (akunszt) [inline bin/bind]: This fails on files with spaces and the `basename` has a second option to trim down the "extension"/matching postfix:
```shell
export CLASS="$(basename "$i" .yaml)"`
```
If you want to solve it with pure Bash then you have to use two lines, you can't nest variable expansions in Bash:
```shell
CLASS="${i##*/}"
export CLASS="${CLASS%.yaml}"
```
Also please use capital I if you're using capital variable names (the I, the ROOT, the CLASS are variables with same visibility in the script, there is no difference).

File: bin/install-mac [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+ROOT="$(cd "$(dirname "$0")/.." && pwd)"
+
+uncrlf() { sed $'s/\r//g'; }
```
Comment (akunszt) [inline bin/install-mac]: It's a little bit simpler:
```shell
tr -d '\r'
```
But you're using this function only once in the middle of a pipeline. I think this can be moved directly to there, it doesn't require to be a shell function.

File: bin/install-mac [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+RELEASES_URL=https://github.com/coreos/container-linux-config-transpiler/releases
+TAG=$(curl -s $RELEASES_URL/latest -D - | uncrlf | grep ^Location: | awk '{ print $2 }' | xargs basename)
+OBJ=ct-$TAG-x86_64-apple-darwin
+BIN_OBJ=$ROOT/bin/$OBJ
+
+curl -Lo $BIN_OBJ $RELEASES_URL/download/$TAG/$OBJ && chmod 755 $BIN_OBJ && ln -sf $BIN_OBJ $ROOT/bin/ct
```
Comment (akunszt) [inline bin/install-mac]: Please quote your variables:
```shell
curl -Lo "$BIN_OBJ" "$RELEASES_URL/download/$TAG/$OBJ" && chmod 755 "$BIN_OBJ" && ln -sf "$BIN_OBJ" "$ROOT/bin/ct"
```
Am I understand correctly that we are start using a new provision tool named `ct`? Why? We are using our own Scala based thing plus a ton of shell scripts to generate CFn/UserData, we are using `Terraform`/`TerraGrunt` for GCP, we are using `Helm` for some Kubernetes deployments. Do we really need a new tool? Or do I misunderstand what is `ct` for?

File: bin/push-to-s3 [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+ROOT="$(cd "$(dirname "$0")/.." && pwd)"
+
+ACCOUNT=hiyautil
+VERSION=$(cat $ROOT/VERSION)
```
Comment (akunszt) [inline bin/push-to-s3]: It's simpler (and faster):
```shell
VERSION="$(<"$ROOT/VERSION")"
```
If you want to stick with the `cat` version then please quote your variables:
```shell
VERSION="$(cat "$ROOT/VERSION")"
```

File: bin/push-to-s3 [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  for region in $push_regions; do
+    bucket=${AWS_S3_BUCKET:-hiya-system-$region}
+    echo "Pushing to $bucket"
+
+    aws-keychain exec $ACCOUNT aws --region $region s3 sync $ROOT/target s3://$bucket/ignition/kafka/$VERSION
+    sleep 0.1 # Give the user a chance to ctrl-C
+  done
+}
+
+DEPLOY_REGION=$1
```
Comment (akunszt) [inline bin/push-to-s3]: Why is this necessary? Why you can't use the `$1` in the next line?

File: bin/push-to-s3 [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+push-to-s3 $DEPLOY_REGION
```
Comment (akunszt) [inline bin/push-to-s3]: Why the function call is necessary? It's a one time thing in a one line script. I think you can paste your function body here, it might easier to read.
Also please quote your variables:
```shell
push-to-s3 "$DEPLOY_REGION"
```

File: bin/push-to-s3 [hunk 10]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+ROOT="$(cd "$(dirname "$0")/.." && pwd)"
+
+ACCOUNT=hiyautil
+VERSION=$(cat $ROOT/VERSION)
+
+push-to-s3() {
+  push_regions="$@"
```
Comment (akunszt) [inline bin/push-to-s3]: Why is necessary? Why do you want to declare a `push_regions` global variable? You're using capital letters in your variable names in the "main code" why is this lowercase?
You can use the `$@` or just the `$1` (as you call this only once with only one variable) in the function too. You don't have to put into a variable. If you have multiple variables and you want to iterate them over then use an array like this:
```shell
PUSH_REGIONS=( "$@" )
for REGION in "${PUSH_REGIONS[@]}"
do
  ...
done
```
It handles the whitespaces while the the non-array version don't. Please note, this is a general advice as in your case you made sure that it will be called with argument always.

File: bin/push-to-s3 [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+  for region in $push_regions; do
```
Comment (akunszt) [inline bin/push-to-s3]: Please use the array based iteration described above or just drop the whole `for` cycle. You don't need it in this case.

File: bin/push-to-s3 [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    bucket=${AWS_S3_BUCKET:-hiya-system-$region}
```
Comment (akunszt) [inline bin/push-to-s3]: This is a global variable as `DEPLOY_REGION`. Please be consequent in your variable names (lowercase XOR uppercase).

File: bin/push-to-s3 [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    echo "Pushing to $bucket"
+
+    aws-keychain exec $ACCOUNT aws --region $region s3 sync $ROOT/target s3://$bucket/ignition/kafka/$VERSION
```
Comment (akunszt) [inline bin/push-to-s3]: Please, quotes:
```shell
aws-keychain exec "$ACCOUNT" aws --region "$region" s3 sync "$ROOT/target" "s3://$bucket/ignition/kafka/$VERSION"
```

File: bin/push-to-s3 [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    sleep 0.1 # Give the user a chance to ctrl-C
```
Comment (akunszt) [inline bin/push-to-s3]: I don't get it why is this necessary. `0.1` second is too fast for human intervention, the Bash can handle the SIGINT without this and as the cycle will run only one time it's already cry over spilt milk.

File: cl-config/kafka-ignition.yaml [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+
+          instance_tags=$(docker run $awscli_image ec2 describe-instances --region $region --instance-ids $instance_id \
+              | jq '.Reservations[].Instances[].Tags | map({(.Key): .Value}) | add')
+
+          instance_name=$(jq -r '.Name' <<< $instance_tags)
+          base_stack=$(jq -r '.BaseStack' <<< $instance_tags)
+          volume_id=$(jq -r '.VolumeId' <<< $instance_tags)
+
+          kafka_heap_size=4g
+          mem_total_kb=$(cat /proc/meminfo | grep MemTotal | awk '{print $2}')
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: You don't need the extra `cat` command and it's worth to use the longest match to avoid unwanted results in the output.
```shell
grep ^MemTotal: /proc/meminfo | ...
```

File: cl-config/kafka-ignition.yaml [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+          if ((mem_total_kb < 8000000)); then
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: It's simpler:
```shell
(( MEM_TOTAL_KB < 8000000 )) && KAFKA_HEAP_SIZE="1g"
```

File: cl-config/kafka-ignition.yaml [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+
+          KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181
+          KAFKA_LISTENERS=PLAINTEXT://0.0.0.0:9092
+          KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://$instance_name:9092
+          KAFKA_NUM_PARTITIONS=6
+          KAFKA_DEFAULT_REPLICATION_FACTOR=3
+          KAFKA_MIN_INSYNC_REPLICAS=2
+          KAFKA_COMPRESSION_TYPE=snappy
+          KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR=3
+          KAFKA_HEAP_OPTS='-Xmx${kafka_heap_size} -Xms${kafka_heap_size} -XX:MetaspaceSize=96m -XX:+UseG1GC -XX:MaxGCPauseMillis=20 -XX:InitiatingHeapOccupancyPercent=35 -XX:G1HeapRegionSize=16M -XX:MinMetaspaceFreeRatio=50 -XX:MaxMetaspaceFreeRatio=80'
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: The variable expansion is not necessary here. You can write `$kafka_heap_size` instead of `${kafka_heap_size}`.

File: cl-config/kafka-ignition.yaml [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+      contents:
+        inline: |
+          #!/usr/bin/env bash
+
+          private_ip=$(curl $METADATA_URL/local-ipv4)
+          hosted_zone=$(docker run $AWS_CLI_IMAGE route53 list-hosted-zones-by-name \
+              | jq -r '.HostedZones[] | select(.Name == "'$BASE_URL.'") | .Id' \
+              | cut -d / -f 3)
+
+          change_batch=$(cat << EOF
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: You don't need a sub-shell and a here-doc here. You can use a multi-line variable easily.
```shell
CHANGE_BATCH="
  {
     ...
  }
"
```

File: cl-config/kafka-ignition.yaml [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+                  "Name": "${INSTANCE_NAME}",
+                  "Type": "A",
+                  "TTL": 60,
+                  "ResourceRecords": [{ "Value": "$private_ip" }]
+                }
+              }]
+            }
+          EOF
+          )
+          docker run $AWS_CLI_IMAGE route53 change-resource-record-sets --hosted-zone-id $hosted_zone --change-batch "$change_batch"
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: Please, quotes:
```shell
docker run "$AWS_CLI_IMAGE" route53 change-resource-record-sets --hosted-zone-id "$hosted_zone" --change-batch "$change_batch"
```

File: cl-config/kafka-ignition.yaml [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+
+    - path: /opt/bin/kafka-attach-volume
+      mode: 0755
+      filesystem: root
+      contents:
+        inline: |
+          #!/usr/bin/env bash
+
+          instance_id=$(curl $METADATA_URL/instance-id)
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: You might want to check the `curl` exit code. Or at least the `instance_id` content (and please quote your variables):
```shell
error()
{
  echo "ERROR: $1" >&2
  exit 1
}


INSTANCE_ID="$(curl "$METADATA_URL/instance-id")"
[[ -z "$INSTANCE_ID" ]] && error "Empty instance id (cURL failed?)!"
```

File: cl-config/kafka-ignition.yaml [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+          volume_id=$VOLUME_ID
+          device='/dev/xvdg'
+
+          docker run $AWS_CLI_IMAGE ec2 detach-volume --region $REGION --volume-id $volume_id
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: Please, quotes:
```shell
docker run "$AWS_CLI_IMAGE" ec2 detach-volume --region "$REGION" --volume-id "$VOLUME_ID"
```

File: cl-config/kafka-ignition.yaml [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+
+    - path: /opt/bin/kafka-attach-volume
+      mode: 0755
+      filesystem: root
+      contents:
+        inline: |
+          #!/usr/bin/env bash
+
+          instance_id=$(curl $METADATA_URL/instance-id)
+          volume_id=$VOLUME_ID
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: Why is this necessary? The uppercase characters won't hurt you and both variables (`volume_id` and `VOLUME_ID`) are global variables.
In a script there is no such thing that environmental variable. The environmental variable is a shell variable set before the script starts but inside the script you can't decide the origin of the variable.

File: cl-config/kafka-ignition.yaml [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+          device='/dev/xvdg'
+
+          docker run $AWS_CLI_IMAGE ec2 detach-volume --region $REGION --volume-id $volume_id
+
+          until [[ "available" == $(docker run $AWS_CLI_IMAGE --region $REGION ec2 describe-volumes --volume-id $volume_id | jq -r '.Volumes[].State') ]]; do
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: Please, quotes:
```bash
until [[ "available" == "$(docker run "$AWS_CLI_IMAGE" --region "$REGION" ec2 describe-volumes --volume-id "$VOLUME_ID" | jq -r '.Volumes[].State')" ]]
```
Also it might to add a counter variable to exit if the volume isn't available at all.
```shell
I=0
while (( I < 120 )) && [[ "$(...)" != "available" ]]
do
  echo "[$I]: Volume is not yet available, waiting..."
  sleep 5
  (( I++ ))
done
(( I == 120 )) && error "Volume is not available! Timeout!"
```

File: cl-config/kafka-ignition.yaml [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+            echo "volume $volume_id is not yet available. waiting..."
+            sleep 5
+          done
+
+          docker run $AWS_CLI_IMAGE ec2 attach-volume --region $REGION --device $device --instance-id $instance_id --volume-id $volume_id
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: Please, quotes:
```shell
docker run "$AWS_CLI_IMAGE" ec2 attach-volume --region "$REGION" --device "$device" --instance-id "$instance_id" --volume-id "$VOLUME_ID"
```

File: cl-config/kafka-ignition.yaml [hunk 25]
```diff
@@ -0,0 +1,367 @@
...
+        [Service]
+        EnvironmentFile=/etc/hiya/kafka.env
+        Environment=KAFKA_LOG_DIRS=/logs
+        Restart=always
+        RestartSec=5
+        ExecStartPre=/bin/mkdir -p /mnt/kafka/data /mnt/kafka/logs
+        ExecStart=/usr/bin/docker run --rm --name kafka --net host \
+          --volume /mnt/kafka/logs:/logs \
+          --volume /mnt/kafka/data:/data \
+          --env KAFKA_LISTENERS=${KAFKA_LISTENERS} \
```
Comment (akunszt) [inline cl-config/kafka-ignition.yaml]: You don't need the variable expansions here, you can use just the variable names, e.g. `--env KAFKA_LISTENERS="$KAFKA_LISTENERS"`

File: cl-config/kafka-zookeeper-ignition.yaml [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+      filesystem: root
+      contents:
+        inline: |
+          #!/usr/bin/env bash
+
+          awscli_image='hiya-public-docker.jfrog.io/platform/awscli:v1.11.120'
+          metadata_url='169.254.169.254/latest/meta-data'
+          instance_id=$(curl $metadata_url/instance-id)
+          region=$(curl $metadata_url/placement/availability-zone | sed -e 's/.$//')
+          base_url=$(curl $metadata_url/hostname | sed -e 's/[^\.]*\.//')
```
Comment (akunszt) [inline cl-config/kafka-zookeeper-ignition.yaml]: As I wrote earlier it might to worth to check the variable's content before going on:
```shell
[[ -z "$INSTANCE_ID" ]] && error "Missing instance id!"
[[ -z "$REGION" ]] && error "No region!"
[[ -z "$BASE_URL" ]] && error "Failed to acquire base URL!"
```
Also please quote the `$metadata_url`.

File: cl-config/kafka-zookeeper-ignition.yaml [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+
+          instance_definition=$(docker run $awscli_image ec2 describe-instances --region $region --instance-ids $instance_id \
```
Comment (akunszt) [inline cl-config/kafka-zookeeper-ignition.yaml]: Please, quotes:
```shell
instance_definition=$(docker run "$awscli_image" ec2 describe-instances --region "$region" --instance-ids "$instance_id" | ...)
```

File: cl-config/kafka-zookeeper-ignition.yaml [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+              | jq '.Reservations[0].Instances[0]')
+
+          instance_name=$(jq -r '.Tags[] | select(.Key == "Name"       ) | .Value' <<< $instance_definition)
+          base_stack=$(   jq -r '.Tags[] | select(.Key == "BaseStack"  ) | .Value' <<< $instance_definition)
+          zookeeper_id=$( jq -r '.Tags[] | select(.Key == "ZookeeperId") | .Value' <<< $instance_definition)
+
+          mkdir -p /etc/hiya
+
+          cat << EOF > /etc/hiya/zookeeper.env
+          REGION=$region
```
Comment (akunszt) [inline cl-config/kafka-zookeeper-ignition.yaml]: Quote your variables. Here you're preparing a file. If the `$region` (or any other variable) has whitespaces then it won't work. It's different than assigning a value directly.
This works (ugly but works):
```shell
MYVAR="text with spaces"
REGION=$MYVAR
```
This won't work:
```shell
REGION=text with spaces
```
In this case you're doing the latter one. You're creating a file with that content which can't be sourced back again. Which is the better result, the worst is that it tries to execute the text after the first space as a command, so it can be considered as a code execution security hole. The attack vector is very small, to manipulate the instance definition already assumes very high permission.

File: cl-config/kafka-zookeeper-ignition.yaml [hunk 29]
```diff
@@ -0,0 +1,221 @@
...
+      mode: 0755
+      filesystem: root
+      contents:
+        inline: |
+          #!/usr/bin/env bash
+
+          awscli_image='hiya-public-docker.jfrog.io/platform/awscli:v1.11.120'
+          metadata_url='169.254.169.254/latest/meta-data'
+
+          private_ip=$(curl $metadata_url/local-ipv4)
```
Comment (akunszt) [inline cl-config/kafka-zookeeper-ignition.yaml]: Please check the variable content and quote the `$metada_url`.

File: cl-config/kafka-zookeeper-ignition.yaml [hunk 30]
```diff
@@ -0,0 +1,221 @@
...
+        EnvironmentFile=/etc/environment
+        EnvironmentFile=/etc/hiya/zookeeper.env
+        Restart=always
+        RestartSec=10
+        # Zk leader election can fail and cause the cluster to get stuck in an infinite startup loop if
+        # the broker's hostname isn't resolvable during initial cluster assembly; ref. error message:
+        # 'As I'm leaving the listener thread, I won't be able to participate in leader election any longer'
+        ExecStartPre=/bin/bash -c '[[ $(host $(hostname -f) | awk "{print \$4}") == $(curl 169.254.169.254/latest/meta-data/local-ipv4) ]]'
+        ExecStartPre=/usr/bin/systemctl is-active docker.service
+        ExecStartPre=-/usr/bin/docker rm zookeeper
```
Comment (akunszt) [inline cl-config/kafka-zookeeper-ignition.yaml]: I'm not familiar with the systemd units, the dash (`-`) has a special meaning here or it's a typo?

**Comment by akunszt** [comment]
I added a ton of simple comments. Please check them. Next time I'll hit the "Start review" button, I didn't know that the simple comments won't be shown here.

File: cl-config/kafka-zookeeper-ignition.yaml (hunk 30 above)
Comment (jutley) [reply cl-config/kafka-zookeeper-ignition.yaml]: The dash means it's okay if the process fails (in this case, the container may not exist, and that's okay).

More here: https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStart=

File: bin/install-mac (hunk 6 above)
Comment (jutley) [reply bin/install-mac]: `ct` (config transpiler) is used to convert CoreOS Container Linux configuration to Ignition configuration (which is platform independent).

Ignition configuration is a replacement for UserData. So the Scala-based CFn generation is still relevant, though Jack is working to update it to use Ignition instead. This change is necessary because our K8s UserData is bumping against the size limit. There are also reasons ignition is a better tool: 
- It runs earlier (during initrd), allowing us to specify behavior before cloud-config would have run (avoiding some race conditions)
- It only runs on first boot, simplifying the startup behavior on future runs
- It allows us to have versioned artifacts for instance configuration in a way that is independent of instance UserData

So yes, it is another tool, but not without reason. And these files I just copied from the aws-templates repo (I wanted a clean repo to work on this Kafka stuff). Jack can tell you more if you are curious/concerned.

File: bin/push-to-s3 (hunk 10 above)
Comment (jutley) [reply bin/push-to-s3]: So, this is not my original code, but I want to get your opinion since I'm finding myself wanted to disagree with you on a couple things. 

So, in your example, I would prefer that REGION does not have a value outside of the for loop. Is this possible?

If not, I'd still like to indicate that REGION is not a constant. One way I might try to indicate this is by making it lowercase. Is there a different convention you might follow to distinguish constants from variables?

Also, I just learned about the readonly keyword. Is that something you would actually use for constants?

File: cl-config/kafka-ignition.yaml (hunk 17 above)
Comment (jutley) [reply cl-config/kafka-ignition.yaml]: I often like to use the variable expansion syntax for the sake of readability. IMO, `-Xmx${kafka_heap_size}` is easier to read than `-Xmx$kafka_heap_size`. Do you think this is not appropriate use of the syntax?

File: cl-config/kafka-ignition.yaml (hunk 18 above)
Comment (jutley) [reply cl-config/kafka-ignition.yaml]: Sure, but then I have to escape every quotation, right? That seems like a much bigger PITA than what I have here.

File: bin/push-to-s3 (hunk 10 above)
Comment (akunszt) [reply bin/push-to-s3]: To answer your questions/concerns:

> So, in your example, I would prefer that REGION does not have a value outside of the for loop. Is this possible?

No. Every variable is a global or a function local variable in Bash. You can't limit the visibility more. You can't have "block local" variables.

> If not, I'd still like to indicate that REGION is not a constant.

It's easy as there is no such thing like constant in Bash. Everything is a variable which you can modify (even the read-only ones in certain Bash versions using namerefs - it was a long standing bug -, don't depend on it).

My convention is to use UPPERCASE variable names and lowercase function names. You can create a variable and a function using the same name but it's a bad pattern as it makes harder to detect typos (e.g. `VAR newvalue`, if is there a `VAR` named function then it's valid but this is not what you probably wanted to do).

> Also, I just learned about the readonly keyword. Is that something you would actually use for constants?

No. I don't use the `readonly` keyword which is actually not a keyword but a builtin command. As there is no such thing is constants in Bash I never wanted to use them. The big trade off of the read-only variables that you can't even unset them. It will pollute your shell until you start a new one.

Beside that I don't know why do you need constants at all? It's not a general interpreted or compiled language, it's a shell. It's good interactive usage and writing short-medium scripts to automate things but it has serious limitations. Some of them you can lift, some of them you can't. In general, if you can't live with a limitation then you're using the wrong tool for the task.

File: cl-config/kafka-ignition.yaml (hunk 17 above)
Comment (akunszt) [reply cl-config/kafka-ignition.yaml]: If it makes the code more readable then you can use the `-Xmx${var}` form too. I might to stare at shell scripts too long as I can read the `-Xmx$var` as easily as the other form.

There isn't any performance penalty and AFAIK they are functionally the same. I just see too often that people thinks that the `${var}` form is the only legit form.

File: cl-config/kafka-ignition.yaml (hunk 18 above)
Comment (akunszt) [reply cl-config/kafka-ignition.yaml]: Yes, you have to escape the string then. It's painful to write but the `cat <<EOF` form is a magnitude slower to execute and it has non-obvious requirements.
```
arpad@trantor ~ $ umask 777
arpad@trantor ~ $ cat <<EOF
> Hello world
> EOF
bash: cannot create temp file for here-document: Permission denied
```
What just happened? The here-doc (`<<`) and the here-string (`<<<`) creates a temporary file. This is why it's not recommended to use them in early startup scripts where the `/tmp` is not yet writable. And they are failing if the user sets the `umask` to funny values also.

File: cl-config/kafka-ignition.yaml (hunk 25 above)
Comment (akunszt) [reply cl-config/kafka-ignition.yaml]: Please let me now the results. I'm not familiar with systemd. I made my recommendations based on that assumption that the systemd is sane and behaves like Bash. I think the systemd is sane pre-condition is a bit too strong...

File: cl-config/kafka-zookeeper-ignition.yaml (hunk 30 above)
Comment (akunszt) [reply cl-config/kafka-zookeeper-ignition.yaml]: Thanks. I didn't know that.
